File: Components/DevExpressVCL/Demos/ExpressQuantumTreeList/CBuilder/EditorsInPlaceValidationDemo/EditorsInPlaceValidationDemoMain.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "EditorsInPlaceValidationDemoMain.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "cxClasses"
#pragma link "cxControls"
#pragma link "cxCustomData"
#pragma link "cxData"
#pragma link "cxDBData"
#pragma link "cxEdit"
#pragma link "cxFilter"
#pragma link "cxGraphics"
#pragma link "cxStyles"
#pragma link "cxLookAndFeels"
#pragma link "DemoBasicMain"
#pragma link "cxButtonEdit"
#pragma link "cxCalc"
#pragma link "cxCalendar"
#pragma link "cxDBLookupComboBox"
#pragma link "cxDBTL"
#pragma link "cxDropDownEdit"
#pragma link "cxImageComboBox"
#pragma link "cxInplaceContainer"
#pragma link "cxSpinEdit"
#pragma link "cxTimeEdit"
#pragma link "cxTL"
#pragma link "cxTLData"
#pragma link "cxBlobEdit"
#pragma link "cxCheckBox"
#pragma link "cxCurrencyEdit"
#pragma link "cxHyperLinkEdit"
#pragma link "cxImage"
#pragma link "cxMaskEdit"
#pragma link "cxMemo"
#pragma link "cxMRUEdit"
#pragma link "cxRadioGroup"
#pragma link "cxTextEdit"
#pragma link "cxLookAndFeelPainters"
#pragma link "cxTLdxBarBuiltInMenu"
#pragma link "cxHint"
#pragma link "dxCustomHint"
#pragma link "dxScreenTip"
#pragma resource "*.dfm"
TEditorsInPlaceValidationDemoMainForm *EditorsInPlaceValidationDemoMainForm;
//---------------------------------------------------------------------------
__fastcall TEditorsInPlaceValidationDemoMainForm::TEditorsInPlaceValidationDemoMainForm(TComponent* Owner)
  : TDemoBasicMainForm(Owner)
{
}
//---------------------------------------------------------------------------

void __fastcall TEditorsInPlaceValidationDemoMainForm::miShowEditBtnsClick(TObject *Sender)
{
  ((TMenuItem*)Sender)->Checked = true;
  cxTreeList1->OptionsView->ShowEditButtons =
    (TcxEditingControlEditShowButtons)((TMenuItem*)Sender)->Tag;
}
//---------------------------------------------------------------------------
void __fastcall TEditorsInPlaceValidationDemoMainForm::cxHintStyleControllerShowHintEx(TObject *Sender,
		  String &Caption, String &HintStr, bool &CanShow, THintInfo &HintInfo)

{
	stGrid->Header->Glyph->Assign(NULL);
	TObject *AViewInfo = static_cast<TObject*>(HintInfo.HintData);
	if (dynamic_cast<TcxTreeListEditCellViewInfo*>(AViewInfo))
	{
		switch ((static_cast<TcxTreeListEditCellViewInfo*>(AViewInfo))->EditViewInfo->ErrorData->ErrorType)
		{
			case eetError: {stGrid->Header->Glyph->Assign(cxEditErrorIcon); break;}
			case eetWarning: {stGrid->Header->Glyph->Assign(cxEditWarningIcon); break; }
			case eetInfo: {stGrid->Header->Glyph->Assign(cxEditInfoIcon); break; }
			case eetCustom: {stGrid->Header->Glyph->Assign(icCustomIcon1->Picture->Bitmap); break; }
		}
	}
}
//---------------------------------------------------------------------------

void __fastcall TEditorsInPlaceValidationDemoMainForm::InitializeEditors(TObject *Sender)

{
  TcxEditValidationOptions AValidationOptions;
  if (miValidationRaiseException->Checked)
  {
	AValidationOptions << evoRaiseException;
  }
  if (miValidationShowErrorIcon->Checked)
  {
	AValidationOptions << evoShowErrorIcon;
  }
  if (miValidationAllowLoseFocus->Checked)
  {
	AValidationOptions << evoAllowLoseFocus;
  }

  cxTreeList1ColumnFirstName->Properties->ValidationOptions = AValidationOptions;
  cxTreeList1ColumnLastName->Properties->ValidationOptions = AValidationOptions;
  cxTreeList1ColumnAddress->Properties->ValidationOptions = AValidationOptions;
  cxTreeList1ColumnPhoneNumber->Properties->ValidationOptions = AValidationOptions;
  cxTreeList1ColumnEmail->Properties->ValidationOptions = AValidationOptions;
}
//---------------------------------------------------------------------------

void __fastcall TEditorsInPlaceValidationDemoMainForm::FormCreate(TObject *Sender)

{
  InitializeEditors(this);
}
//---------------------------------------------------------------------------
bool TEditorsInPlaceValidationDemoMainForm::DoAddressValidate(const Variant &AValue, TCaption &AErrorText)
{
  int I = (dynamic_cast<TcxComboBoxProperties*> (cxTreeList1ColumnAddress->Properties))->Items->IndexOf(VarToStr(AValue));
  bool rv = (I == -1);
  if (rv)
  {
	AErrorText = "Please select an address from the list";
  }
  return rv;
}
//---------------------------------------------------------------------------

bool TEditorsInPlaceValidationDemoMainForm::DoEmailValidate(const Variant &AValue, TCaption &AErrorText)
{
  String S = VarToStr(AValue);
  bool rv = ((S != "")&&(!Cxregexpr::IsTextFullValid(S, "[A-z0-9_-]+@[A-z0-9_-]+\\.[A-z0-9_-]+(\\.[A-z]+)*")));
  if (rv)
  {
	AErrorText = "Please enter a valid email address";
  }
  return rv;
}
//---------------------------------------------------------------------------

bool TEditorsInPlaceValidationDemoMainForm::DoFirstNameValidate(const Variant &AValue, TCaption &AErrorText)
{
  bool rv = (VarToStr(AValue) == "");
  if (rv)
  {
	AErrorText = "Please enter a value";
  }
  return rv;
}
//---------------------------------------------------------------------------

bool TEditorsInPlaceValidationDemoMainForm::DoLastNameValidate(const Variant &AValue, TCaption &AErrorText)
{
  bool rv = (VarToStr(AValue) == "");
  if (rv)
  {
	AErrorText = "Please enter a value";
  }
  return rv;
}
//---------------------------------------------------------------------------

bool TEditorsInPlaceValidationDemoMainForm::DoPhoneNumberValidate(const Variant &AValue, TCaption &AErrorText)
{
  String S = VarToStr(AValue);
  bool rv = ((S == "")||(!Cxregexpr::IsTextValid(S, "(\\(\\d\\d\\d\\)' ')?\\d\\d\\d-\\d\\d\\d\\d")));
  if (rv)
  {
	AErrorText = "Please enter a valid phone number";
  }
  return rv;
}
void __fastcall TEditorsInPlaceValidationDemoMainForm::cxTreeList1ColumnFirstNamePropertiesValidate(TObject *Sender,
          Variant &DisplayValue, TCaption &ErrorText, bool &Error)
{
  Error = DoFirstNameValidate(DisplayValue, ErrorText);
}
//---------------------------------------------------------------------------
void __fastcall TEditorsInPlaceValidationDemoMainForm::cxTreeList1ColumnFirstNameValidateDrawValue(TcxTreeListColumn *Sender,
          TcxTreeListNode *ANode, const Variant &AValue, TcxEditValidateInfo *AData)
{
  String AErrorText;
  if (DoFirstNameValidate(AValue, AErrorText))
  {
	AData->ErrorType = eetError;
	AData->ErrorText = AErrorText;
  }
}
//---------------------------------------------------------------------------
void __fastcall TEditorsInPlaceValidationDemoMainForm::cxTreeList1ColumnAddressPropertiesValidate(TObject *Sender,
          Variant &DisplayValue, TCaption &ErrorText, bool &Error)
{
  Error = DoAddressValidate(DisplayValue, ErrorText);
}
//---------------------------------------------------------------------------
void __fastcall TEditorsInPlaceValidationDemoMainForm::cxTreeList1ColumnAddressValidateDrawValue(TcxTreeListColumn *Sender,
          TcxTreeListNode *ANode, const Variant &AValue, TcxEditValidateInfo *AData)
{
  String AErrorText;
  if (DoAddressValidate(AValue, AErrorText))
  {
	AData->ErrorType = eetInfo;
	AData->ErrorText = AErrorText;
  }
}
//---------------------------------------------------------------------------
void __fastcall TEditorsInPlaceValidationDemoMainForm::cxTreeList1ColumnEmailPropertiesValidate(TObject *Sender,
          Variant &DisplayValue, TCaption &ErrorText, bool &Error)
{
  Error = DoEmailValidate(DisplayValue, ErrorText);
}
//---------------------------------------------------------------------------
void __fastcall TEditorsInPlaceValidationDemoMainForm::cxTreeList1ColumnEmailValidateDrawValue(TcxTreeListColumn *Sender,
          TcxTreeListNode *ANode, const Variant &AValue, TcxEditValidateInfo *AData)
{
  String AErrorText;
  if (DoEmailValidate(AValue, AErrorText))
  {
	AData->ErrorType = eetCustom;
	AData->ErrorIcon->Assign(icCustomIcon1->Picture->Bitmap);
	AData->ErrorText = AErrorText;
  }

}
//---------------------------------------------------------------------------
void __fastcall TEditorsInPlaceValidationDemoMainForm::cxTreeList1ColumnLastNamePropertiesValidate(TObject *Sender,
          Variant &DisplayValue, TCaption &ErrorText, bool &Error)
{
  Error = DoLastNameValidate(DisplayValue, ErrorText);
}
//---------------------------------------------------------------------------
void __fastcall TEditorsInPlaceValidationDemoMainForm::cxTreeList1ColumnLastNameValidateDrawValue(TcxTreeListColumn *Sender,
          TcxTreeListNode *ANode, const Variant &AValue, TcxEditValidateInfo *AData)

{
  String AErrorText;
  if (DoLastNameValidate(AValue, AErrorText))
  {
	AData->ErrorType = eetError;
	AData->ErrorText = AErrorText;
  }
}
//---------------------------------------------------------------------------
void __fastcall TEditorsInPlaceValidationDemoMainForm::cxTreeList1ColumnPhoneNumberPropertiesValidate(TObject *Sender,
          Variant &DisplayValue, TCaption &ErrorText, bool &Error)
{
  Error = DoPhoneNumberValidate(DisplayValue, ErrorText);
}
//---------------------------------------------------------------------------
void __fastcall TEditorsInPlaceValidationDemoMainForm::cxTreeList1ColumnPhoneNumberValidateDrawValue(TcxTreeListColumn *Sender,
          TcxTreeListNode *ANode, const Variant &AValue, TcxEditValidateInfo *AData)
{
  String AErrorText;
  if (DoPhoneNumberValidate(AValue, AErrorText))
  {
	AData->ErrorType = eetError;
	AData->ErrorText = AErrorText;
  }
}
//---------------------------------------------------------------------------

File: Components/DevExpressVCL/Demos/ExpressQuantumTreeList/Delphi/EditorsInPlaceValidationDemo/EditorsInPlaceValidationDemoMain.pas
unit EditorsInPlaceValidationDemoMain;

{$I cxVer.inc}

interface

uses
  Windows, Messages, SysUtils, Variants, Classes, Graphics, Forms, Controls,
  Dialogs, cxControls, cxLookAndFeels, ActnList, ImgList, Menus, ComCtrls,
  StdCtrls,
{$IFDEF EXPRESSBARS}
  dxBar, dxStatusBar,
{$ENDIF}
  DemoBasicMain,
  cxContainer, cxEdit, cxTextEdit, cxStyles, cxTL,
  cxMaskEdit, cxCalendar, cxCurrencyEdit, cxMemo, cxInplaceContainer,
  cxDBTL, cxTLData, cxDBLookupComboBox, cxTimeEdit, cxCalc, cxSpinEdit,
  cxImageComboBox, cxDropDownEdit, cxEditRepositoryItems, cxRadioGroup,
  cxImage, cxBlobEdit, cxCheckBox, cxHyperLinkEdit, cxButtonEdit, cxMRUEdit,
  cxGraphics, cxCustomData, cxDataUtils, cxLookAndFeelPainters, cxTLdxBarBuiltInMenu,
  cxRegExpr, cxClasses, dxScreenTip, dxCustomHint, cxHint;

type
  TEditorsInPlaceValidationDemoMainForm = class(TDemoBasicMainForm)
    miValidationOptions: TMenuItem;
    miValidationRaiseException: TMenuItem;
    miValidationShowErrorIcon: TMenuItem;
    miValidationAllowLoseFocus: TMenuItem;
    N1: TMenuItem;
    cxTreeList1: TcxTreeList;
    cxTreeList1ColumnFirstName: TcxTreeListColumn;
    cxTreeList1ColumnLastName: TcxTreeListColumn;
    cxTreeList1ColumnAddress: TcxTreeListColumn;
    cxTreeList1ColumnPhoneNumber: TcxTreeListColumn;
    cxTreeList1ColumnEmail: TcxTreeListColumn;
    icCustomIconList: TcxImageCollection;
    icCustomIcon1: TcxImageCollectionItem;
    cxHintStyleController: TcxHintStyleController;
    dxScreenTipRepository: TdxScreenTipRepository;
    stGrid: TdxScreenTip;
    procedure cxTreeList1ColumnFirstNameValidateDrawValue(
      Sender: TcxTreeListColumn; ANode: TcxTreeListNode;
      const AValue: Variant; AData: TcxEditValidateInfo);
    procedure cxTreeList1ColumnFirstNamePropertiesValidate(Sender: TObject;
      var DisplayValue: Variant; var ErrorText: TCaption;
      var Error: Boolean);
    procedure cxTreeList1ColumnLastNameValidateDrawValue(
      Sender: TcxTreeListColumn; ANode: TcxTreeListNode;
      const AValue: Variant; AData: TcxEditValidateInfo);
    procedure cxTreeList1ColumnLastNamePropertiesValidate(Sender: TObject;
      var DisplayValue: Variant; var ErrorText: TCaption;
      var Error: Boolean);
    procedure cxTreeList1ColumnAddressValidateDrawValue(
      Sender: TcxTreeListColumn; ANode: TcxTreeListNode;
      const AValue: Variant; AData: TcxEditValidateInfo);
    procedure cxTreeList1ColumnAddressPropertiesValidate(Sender: TObject;
      var DisplayValue: Variant; var ErrorText: TCaption;
      var Error: Boolean);
    procedure cxTreeList1ColumnPhoneNumberValidateDrawValue(
      Sender: TcxTreeListColumn; ANode: TcxTreeListNode;
      const AValue: Variant; AData: TcxEditValidateInfo);
    procedure cxTreeList1ColumnPhoneNumberPropertiesValidate(
      Sender: TObject; var DisplayValue: Variant; var ErrorText: TCaption;
      var Error: Boolean);
    procedure cxTreeList1ColumnEmailValidateDrawValue(
      Sender: TcxTreeListColumn; ANode: TcxTreeListNode;
      const AValue: Variant; AData: TcxEditValidateInfo);
    procedure cxTreeList1ColumnEmailPropertiesValidate(Sender: TObject;
      var DisplayValue: Variant; var ErrorText: TCaption;
      var Error: Boolean);
    procedure FormCreate(Sender: TObject);
    procedure cxHintStyleControllerShowHintEx(Sender: TObject; var Caption,
      HintStr: string; var CanShow: Boolean; var HintInfo: THintInfo);
  private
    function DoAddressValidate(const AValue: Variant; var AErrorText: TCaption): Boolean;
    function DoEmailValidate(const AValue: Variant; var AErrorText: TCaption): Boolean;
    function DoFirstNameValidate(const AValue: Variant; var AErrorText: TCaption): Boolean;
    function DoLastNameValidate(const AValue: Variant; var AErrorText: TCaption): Boolean;
    function DoPhoneNumberValidate(const AValue: Variant; var AErrorText: TCaption): Boolean;
  published
    procedure InitializeEditors(Sender: TObject);
  end;

var
  EditorsInPlaceValidationDemoMainForm: TEditorsInPlaceValidationDemoMainForm;

implementation

uses
  ShellAPI;

{$R *.dfm}

function GetMenuItemChecked(AMenuItem: TObject): Boolean;
begin
{$IFDEF EXPRESSBARS}
  if AMenuItem is TdxBarButton then
    Result := TdxBarButton(AMenuItem).Down
  else
{$ENDIF}
   Result := (AMenuItem as TMenuItem).Checked;
end;

procedure TEditorsInPlaceValidationDemoMainForm.InitializeEditors(Sender: TObject);
var
  AValidationOptions: TcxEditValidationOptions;
begin
  AValidationOptions := [];
  if GetMenuItemChecked(miValidationRaiseException) then
    Include(AValidationOptions, evoRaiseException);
  if GetMenuItemChecked(miValidationShowErrorIcon) then
    Include(AValidationOptions, evoShowErrorIcon);
  if GetMenuItemChecked(miValidationAllowLoseFocus) then
    Include(AValidationOptions, evoAllowLoseFocus);

  cxTreeList1ColumnFirstName.Properties.ValidationOptions := AValidationOptions;
  cxTreeList1ColumnLastName.Properties.ValidationOptions := AValidationOptions;
  cxTreeList1ColumnAddress.Properties.ValidationOptions := AValidationOptions;
  cxTreeList1ColumnPhoneNumber.Properties.ValidationOptions := AValidationOptions;
  cxTreeList1ColumnEmail.Properties.ValidationOptions := AValidationOptions;
end;

function TEditorsInPlaceValidationDemoMainForm.DoAddressValidate(const AValue: Variant; var AErrorText: TCaption): Boolean;
begin
  Result := VarToStr(AValue) = '';
  if Result then
    AErrorText := 'Please select an address from the list';
end;

function TEditorsInPlaceValidationDemoMainForm.DoEmailValidate(const AValue: Variant;
  var AErrorText: TCaption): Boolean;
var
  S: string;
begin
  S := VarToStr(AValue);
  Result := (S <> '') and not IsTextFullValid(S, '[A-z0-9_-]+@[A-z0-9_-]+\.[A-z0-9_-]+(\.[A-z]+)*');
  if Result then
    AErrorText := 'Please enter a valid email';
end;

function TEditorsInPlaceValidationDemoMainForm.DoFirstNameValidate(const AValue: Variant; var AErrorText: TCaption): Boolean;
begin
  Result := VarToStr(AValue) = '';
  if Result then
    AErrorText := 'Please enter a value';
end;

function TEditorsInPlaceValidationDemoMainForm.DoLastNameValidate(const AValue: Variant;
  var AErrorText: TCaption): Boolean;
begin
  Result := VarToStr(AValue) = '';
  if Result then
    AErrorText := 'Please enter a value';
end;

function TEditorsInPlaceValidationDemoMainForm.DoPhoneNumberValidate(const AValue: Variant; var AErrorText: TCaption): Boolean;
var
  S: string;
begin
  S := VarToStr(AValue);
  Result := (S = '') or not IsTextValid(S, '(\(\d\d\d\)'' '')?\d\d\d-\d\d\d\d');
  if Result then
    AErrorText := 'Please enter a valid phone number';
end;

procedure TEditorsInPlaceValidationDemoMainForm.cxTreeList1ColumnFirstNameValidateDrawValue(
  Sender: TcxTreeListColumn; ANode: TcxTreeListNode; const AValue: Variant;
  AData: TcxEditValidateInfo);
var
  AErrorText: TCaption;
begin
  if DoFirstNameValidate(AValue, AErrorText) then
  begin
    AData.ErrorType := eetError;
    AData.ErrorText := AErrorText;
  end;
end;

procedure TEditorsInPlaceValidationDemoMainForm.cxTreeList1ColumnFirstNamePropertiesValidate(
  Sender: TObject; var DisplayValue: Variant; var ErrorText: TCaption;
  var Error: Boolean);
begin
  Error := DoFirstNameValidate(DisplayValue, ErrorText);
end;

procedure TEditorsInPlaceValidationDemoMainForm.cxTreeList1ColumnLastNameValidateDrawValue(
  Sender: TcxTreeListColumn; ANode: TcxTreeListNode; const AValue: Variant;
  AData: TcxEditValidateInfo);
var
  AErrorText: TCaption;
begin
  if DoLastNameValidate(AValue, AErrorText) then
  begin
    AData.ErrorType := eetError;
    AData.ErrorText := AErrorText;
  end;
end;

procedure TEditorsInPlaceValidationDemoMainForm.cxTreeList1ColumnLastNamePropertiesValidate(
  Sender: TObject; var DisplayValue: Variant; var ErrorText: TCaption;
  var Error: Boolean);
begin
  Error := DoLastNameValidate(DisplayValue, ErrorText);
end;

procedure TEditorsInPlaceValidationDemoMainForm.cxTreeList1ColumnAddressValidateDrawValue(
  Sender: TcxTreeListColumn; ANode: TcxTreeListNode; const AValue: Variant;
  AData: TcxEditValidateInfo);
var
  AErrorText: TCaption;
begin
  if DoAddressValidate(AValue, AErrorText) then
  begin
    AData.ErrorType := eetInfo;
    AData.ErrorText := AErrorText;
  end;
end;

procedure TEditorsInPlaceValidationDemoMainForm.cxHintStyleControllerShowHintEx(
  Sender: TObject; var Caption, HintStr: string; var CanShow: Boolean;
  var HintInfo: THintInfo);
begin
  stGrid.Header.Glyph.Assign(nil);
  if miValidationShowErrorIcon.Checked and (TObject(HintInfo.HintData) is TcxTreeListEditCellViewInfo) then
    case TcxTreeListEditCellViewInfo(HintInfo.HintData).EditViewInfo.ErrorData.ErrorType of
      eetError:
        stGrid.Header.Glyph.Assign(cxEditErrorIcon);
      eetWarning:
        stGrid.Header.Glyph.Assign(cxEditWarningIcon);
      eetInfo:
        stGrid.Header.Glyph.Assign(cxEditInfoIcon);
      eetCustom:
        stGrid.Header.Glyph.Assign(icCustomIcon1.Picture.Bitmap);
    end;
end;

procedure TEditorsInPlaceValidationDemoMainForm.cxTreeList1ColumnAddressPropertiesValidate(
  Sender: TObject; var DisplayValue: Variant; var ErrorText: TCaption;
  var Error: Boolean);
begin
  Error := DoAddressValidate(DisplayValue, ErrorText);
end;

procedure TEditorsInPlaceValidationDemoMainForm.cxTreeList1ColumnPhoneNumberValidateDrawValue(
  Sender: TcxTreeListColumn; ANode: TcxTreeListNode; const AValue: Variant;
  AData: TcxEditValidateInfo);
var
  AErrorText: TCaption;
begin
  if DoPhoneNumberValidate(AValue, AErrorText) then
  begin
    AData.ErrorType := eetWarning;
    AData.ErrorText := AErrorText;
  end;
end;

procedure TEditorsInPlaceValidationDemoMainForm.cxTreeList1ColumnPhoneNumberPropertiesValidate(
  Sender: TObject; var DisplayValue: Variant; var ErrorText: TCaption;
  var Error: Boolean);
begin
  Error := DoPhoneNumberValidate(DisplayValue, ErrorText);
end;

procedure TEditorsInPlaceValidationDemoMainForm.cxTreeList1ColumnEmailValidateDrawValue(
  Sender: TcxTreeListColumn; ANode: TcxTreeListNode; const AValue: Variant;
  AData: TcxEditValidateInfo);
var
  AErrorText: TCaption;
begin
  if DoEmailValidate(AValue, AErrorText) then
  begin
    AData.ErrorType := eetCustom;
    AData.ErrorIcon.Assign(icCustomIconList.Items[0].Picture.Bitmap);
    AData.ErrorText := AErrorText;
  end;
end;

procedure TEditorsInPlaceValidationDemoMainForm.cxTreeList1ColumnEmailPropertiesValidate(
  Sender: TObject; var DisplayValue: Variant; var ErrorText: TCaption;
  var Error: Boolean);
begin
  Error := DoEmailValidate(DisplayValue, ErrorText);
end;

procedure TEditorsInPlaceValidationDemoMainForm.FormCreate(
  Sender: TObject);
begin
  inherited;
  InitializeEditors(Self);
end;

end.
